### PR TITLE
#184: PerpetualStream implementation, tests, and docs

### PR DESCRIPTION
**Pull request by akara**
Thanks for your pull request.  Please review the following guidelines.

- [X] Title includes issue id.
- [X] Description of the change added.
- [X] Commits are squashed.
- [X] Tests added.
- [X] Documentation added/updated.
- [X] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by akara** [comment]
This PR supersedes #203.

File: docs/streams-lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
+# Streams Lifecycle
+Akka Streams/Reactive stream needs to be integrated with runtime lifecycle of the server. For this, an automated or semi-automated integration is provided through `PerpetualStream`. For fine-grained control over stream sources directly, `LifecycleManaged` provides a wrapping that can control any source component for proper stopping and shutdown so the flow can start/stop gracefully.
```
Comment (anilgursel) [inline docs/streams-lifecycle.md]: Should add a link in the first sentence, something like "... with runtime lifecycle of the server (please see [Runtime Lifecycle & API](lifecycle.md)) ."

File: docs/streams-lifecycle.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    // Do all your cleanup
+    // For safety, call super
+    super.shutdown()
+    // The Future from super.shutdown may not mean anything.
+    // Feel free to create your own future that identifies the
+    // stream being done. Return your Future instead.
+  }
+```
+
+`shutdown` needs to do the followings:
```
Comment (anilgursel) [inline docs/streams-lifecycle.md]: followings -> following

File: docs/streams-lifecycle.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+```
+
+**Java**
+
+```java
+import static org.squbs.stream.TriggerEvent.DISABLE;
+import static org.squbs.stream.TriggerEvent.ENABLE;
+
+    final Source<?, ?> inSource = <your-original-source>;
```
Comment (anilgursel) [inline docs/streams-lifecycle.md]: Indentation.

File: docs/streams-lifecycle.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+```
+
+## Custom Lifecycle Event(s) for Trigger
+If you want to respond to more lifecycle events beyond `Active` and `Stopping`, for example you want `Failed` to also stop the flow, you can modify the lifecylce event mapping.
+
+```scala
+import org.squbs.stream.TriggerEvent._
+
+  val inSource = <your-original-source>
```
Comment (anilgursel) [inline docs/streams-lifecycle.md]: Indentation.

Comment (akara) [inline docs/streams-lifecycle.md]: I think I need to be clear what indentation we want in these cases. The `import` statements are generally placed at the beginning of the file for Scala. For Java, they are always placed at the beginning. The rest is code which is in some class. I can do a `...` between the import and the code to make it clearer. But the code indentation is always further in than the `import` statements, except for the Scala case of inline imports. Please let me know how you want to see it. I've addressed all other comments pending sync. Thanks!

Comment (anilgursel) [inline docs/streams-lifecycle.md]: We need to do one of the following:
* add `class Sample {`
* Just indent at the same level with `import` if there is no `class Sample` etc.

I lean towards the second option.  I understand the current indentation has a subtle consideration; however, it looks more of a problem and causes distraction during reading.

Comment (akara) [inline docs/streams-lifecycle.md]: Thanks, will go with option 2.

**Pull request merged**